---
title: "TARGET_AML_FlowSortedSamples_ExploratoryAnalysis.Rmd"
author: "Jenny Smith"
date: "April 2018"
output: html_document
---

```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.height = 8, fig.width = 10)
knitr::opts_knit$set(root.dir = '/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/analysis/2017.10.09_Concatenate_1031_RNAseq/')
```


```{r}
library(tximport)
library(edgeR)
library(limma)
library(RColorBrewer)
source("~/scripts/RNAseq_Analysis/DifferentialExpn_PathwayAnalysis/ggplot_Themes_Function.r")
```


#Examine PCA

```{r}
filepath.1031 <-  "/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/level3/transcript/2017July_BCCA_1031_Kallisto_Illumina_data/"

allfiles <- paste(filepath.1031,dir(path = filepath.1031, pattern = "^.+GRCh37.87.+.tsv$"),sep="")


head(allfiles)
length(allfiles)
```

```{r}
#Sorted primary samples
sorted <- grep("Sorted", allfiles, ignore.case = TRUE, value=TRUE)

```


```{r}
#checking on Bulk. They are not in TARGET. Instead have "Unsorted" in thier identifier. 
str_split_fixed(sorted, "-", n=5)[,3] %>%
  unique(.) %>%
  paste0("^.+", ., ".+") %>%
  paste(., collapse = "|") %>%
  grep(., target)
```


```{r}
sorted.updated <- read.csv("~/RNA_seq_Analysis/0000.00.03_Expression_Matrices/TARGET_AML_AAML1031_Kallisto_Transcript_w.Gene.IDs_RNASeq_FlowSorted_tpm.csv",
                           stringsAsFactors = FALSE)

head(sorted.updated[,1:5])
```


```{r}
plates_qual <- read.csv("~/reference_mapping-files/TARGET_AML_1031_Ribodepleted_RNA_Seq_Rare_Variant_list_with_Quality_Annotated.csv", 
                    stringsAsFactors = FALSE)

# plates_qual$USI <- str_split_fixed(plates_qual$External_ID, "-", n=4)[,3]

plates_qual <- plates_qual %>%
  # select(External_ID,Column1) %>%
  # filter(grepl("sorted", External_ID, ignore.case = TRUE)) %>%
  mutate(Patient_ID=gsub( "(^.+Sorted-non-leukemic-09A)-01R","\\1" ,gsub("TARGET-20-", "",External_ID)),
         USI=str_split_fixed(plates_qual$External_ID, "-", n=4)[,3]) %>%
  mutate(Patient_ID=gsub(" ","-", Patient_ID))

# head(plates_qual)
```

```{r}
#Use Column called RNA to send (ng)
plates_w.Conc <- read.csv("~/reference_mapping-files/TARGET_AML_1031_Rare_variants_for_sequencing_final_v2.csv", 
                    stringsAsFactors = FALSE)

plates_w.Conc <- plates_w.Conc %>% 
  mutate(Patient_ID=gsub( "(^.+Sorted-non-leukemic-09A)-01R","\\1" ,gsub("TARGET-20-", "", plates_w.Conc$X.1))) %>%
  mutate(Patient_ID=gsub(" ","-", Patient_ID)) %>%
  mutate(Patient_ID=gsub("(^.+Sorted-CD34-Myeloid-Mon).+", "\\1",Patient_ID)) %>%
  mutate(Patient_ID=gsub("(^.+Sorted-CD34-Lymphoid-09).+", "\\1", Patient_ID))

```


Note: importing `abundance.h5` is typically faster than `abundance.tsv`
reading in files with read_tsv

https://wurmlab.github.io/genomicscourse/2016-SIB/practicals/rnaseq/TP2
For this purpose, they introduced the "scaledTPM" values, which are obtained by summing the transcript-level TPMs by gene, and multiplying them with the total library size in millions. 

ScaledTPM values are artificial values, transforming underlying abundance measures to the scale of read counts. This allows to incorporate the information provided by the sequencing depth, and work with RNA-seq differential expression tools that were developed to use read counts. 


Jenny, can you do me a favor and 
1.  do two different plots, one for unsorted/blast/no blast sorting and second for unsorted/CD34+ blast/CD34-lymphoid
2.  Add the RNA quantity to each of the dots on the plots

```{r}
tx2gene.df <- select(sorted.updated, transcript_id_version, gene_id)

txi <- tximport(files = sorted, 
                type="kallisto", 
                tx2gene = tx2gene.df, 
                countsFromAbundance = "scaledTPM")

# str(txi)
#Add column names 
colnames(txi$counts) <- gsub(pattern.sorted,"\\1", sorted) #same order as files 
```


```{r}
info <- data.frame(Patient_ID=colnames(txi$counts), stringsAsFactors = FALSE) %>%
 
  mutate(Group = gsub( "^P.+-([SU].+)", "\\1", Patient_ID) %>% 
           gsub("(^[SU].+)-09.{0,}", "\\1", . )) %>%
  mutate(USI=str_split_fixed(Patient_ID,"-",n=2)[,1]) %>%
  
  left_join(., select(plates_w.Conc, Patient_ID, RNA.to.send..ng.), by=c("Patient_ID")) %>%
  left_join(., select(plates_qual, Patient_ID, BCCA.Input.Conc.FAIL=Column1), by=c("Patient_ID")) %>%
  mutate(BCCA.Input.Conc.FAIL=ifelse(grepl("^$", BCCA.Input.Conc.FAIL), NA, BCCA.Input.Conc.FAIL)) %>%
  
  group_by(USI) %>%
  mutate(Flow.Sort.Strategy=ifelse(all(grepl("Sorted-non-leukemic|Unsorted|Sorted-leukemic", Group)), 
                      "Batch1", "Batch2")) %>%
  ungroup() 

head(info)
# write.csv(info, "TARGET_AML_1031_FlowSorted_Groups_info.csv", row.names = FALSE)
# sum(is.na(info$RNA.to.send..ng.))
# sum(is.na(info$Column1))
```


```{r}
dge <- DGEList(txi$counts, samples = info )

keep.dge <- rowSums(cpm(dge) >= 1) > (0.05*ncol(txi$counts)) 
dge <- dge[keep.dge,] #subset for those genes with cmp >= 1 per gene in  samples
dge <- calcNormFactors(dge) #Do TMM normalization

dge <- calcNormFactors(dge)

head(dge$samples)
# dim(dge) # 21873    54
# range(dge$samples$norm.factors) #0.6794583 1.5793262
# range(dge$samples$lib.size) #8,198,930 to 44,539,346
```


```{r}
TMMCPM <- cpm(dge, log=FALSE)

# dim(TMMCPM)
# head(TMMCPM)
```


```{r}
source("~/scripts/RNAseq_Analysis/DifferentialExpn_PathwayAnalysis/rmDupGenes_Function.r")
```


```{r}
Kallisto_IDmap <- read.csv("Homo_sapiens_Ensembl_v75_v76_v87_Transcript.Gene.IDmap.csv",
                           stringsAsFactors = FALSE)

gene_IDmap <- Kallisto_IDmap %>%
  select(gene_id, gene_name) %>%
  unique(.) %>% 
  #picking the first one, because they are all the same gene in ensembl anyway. 
  filter(!duplicated(gene_id))
  # arrange(gene_id)

head(gene_IDmap)
# dim(gene_IDmap) # 38057  genes
```


```{r}
# cts <- list(TMMCPM=TMMCPM, TMMCPM.unfilt=TMMCPM.unfilt)
TMMCPM.Anno <- TMMCPM %>%
  as.data.frame() %>%
  rownames_to_column("Gene") %>%
  inner_join(., gene_IDmap, by=c("Gene"="gene_id")) %>% #there are 1,245 duplicate gene names. will filter. 
  select(geneSymbol=gene_name,everything(),-Gene)

head(TMMCPM.Anno[,1:5])
```


```{r}
rmDupCPM <- rmDupGenes(expnData = TMMCPM.Anno, geneCol = "geneSymbol")

dim(rmDupCPM$remDups) # 20628    54
head(rmDupCPM$remDups[,1:5])
```


```{r}
TMMCPM.Anno.rmDup <- rmDupCPM$remDups %>%
  rownames_to_column("geneSymbol") %>%
  arrange(geneSymbol) %>%
  gather(var,val, -geneSymbol) %>%
  spread(geneSymbol,val) %>% 
  inner_join(., select(info, Patient_ID, Group, USI,Flow.Sort.Strategy), by=c("var"="Patient_ID")) %>%
  select(Patient_ID=var, USI,Group,Flow.Sort.Strategy, everything())

head(TMMCPM.Anno.rmDup[,1:5])
# dim(TMMCPM.Anno) # 54 20632
```


```{r}
A.M <- paste(paste0("^", LETTERS[1:13]), collapse = "|")
N.Z <- paste(paste0("^", LETTERS[14:26]), collapse = "|")

CPM.A.M <- TMMCPM.Anno.rmDup %>%
  select(Patient_ID, USI,Group,Flow.Sort.Strategy, matches(A.M))

CPM.N.Z <- TMMCPM.Anno.rmDup %>%
  select(Patient_ID, USI,Group,Flow.Sort.Strategy, matches(N.Z))
#          
# write.csv(CPM.A.M, "TARGET_AML_1031_FlowSortedSamples_GeneLevelCounts_A-M_fromKallisto_CPM.csv", row.names = FALSE)
# write.csv(CPM.N.Z, "TARGET_AML_1031_FlowSortedSamples_GeneLevelCounts_N-Z_fromKallisto_CPM.csv", row.names = FALSE)
```



```{r}
pca_custom <- function(expnData,CDE){
  
  expnData <- expnData[,intersect(CDE$Patient_ID,colnames(expnData))]
  
  # print(dim(expnData))
  pca <- prcomp(t(expnData), scale=TRUE)
  summ <- summary(pca)

  scores <- as.data.frame(pca$x) %>%
    rownames_to_column("Patient_ID") %>%
    inner_join(., CDE, by="Patient_ID") %>%
    select(Patient_ID, USI, Group, Flow.Sort.Strategy, everything())

  pca.plot <- ggplot(scores, aes(x=PC1, y=PC2)) +
    geom_point(size=6, stroke=4, alpha=0.85,shape=21,
               aes(fill=RNA.to.send..ng., color=Group)) +
    scale_color_manual(values=c("#E41A1C", "#377EB8", "#4DAF4A", "#984EA3", "#FF7F00","#A65628")) +
    labs(x=paste("PC1: ", round(summ$importance[2,1], digits=3)*100, "% variance explained"),
         y=paste("PC2: ", round(summ$importance[2,2], digits=3)*100, "% variance explained")) +
    theme_numX  +
    theme(legend.text = element_text(size=14),
          legend.title = element_text(size=16))

    res <- list(pca=pca,scores=scores,plot=pca.plot)
    return(res)
  
}
```


```{r}
Log2.CPM <- cpm(dge, log = TRUE, prior.count = 1)
dim(Log2.CPM)

head(Log2.CPM[,1:5])
```

```{r fig.width=10, fig.height=5}
PCA.RNA.Conc <- pca_custom(expnData = Log2.CPM, CDE = info) 

```

```{r}

loadings <- as.data.frame(PCA.RNA.Conc$pca$rotation) 

loadings
```

```{r}
loadings[order(abs(loadings$PC1), decreasing = TRUE),]
```


```{r}
head(PCA.RNA.Conc$pca$x[,1:5])
dim(PCA.RNA.Conc$pca$x)
```


```{r}
# as.data.frame(table(info$RNA.to.send..ng., info$Flow.Sort.Strategy)) %>%
#   select(Sample.RNA.ng=Var1,Flow.Sort.Strategy=Var2, everything()) %>% 
#   spread(Flow.Sort.Strategy, Freq) %>%
#   write.csv(., "TARGET_AML_1031_RNA.ng_FrequencyTable.csv", row.names = FALSE)

# as.data.frame(table(info$Flow.Sort.Strategy, info$BCCA.Input.Conc.FAIL)) %>%
#   select(Flow.Sort.Strategy=Var1,BCCA.Input.Conc.FAIL=Var2,Freq) %>%
#   write.csv(., "TARGET_AML_1031_BCCA.Labeled.As.Input.Conc.Fail_FrequencyTable.csv", row.names = FALSE)
```

```{r fig.height=8, fig.width=12}
# tiff("TARGET_AML_1031_FlowSorted_AllSamples_withRNAConc_PCAplot.tiff", res=200, units = "in", height = 8,
#      width = 12)

PCA.RNA.Conc$plot + 
  scale_fill_gradientn(colors=c("lightblue","grey","black"), breaks=c(50,200,300)) + 
  labs(title="PCA Analysis of Flow Sorted AML Samples")

# dev.off()
```

```{r fig.height=8, fig.width=12, message=FALSE, warning=FALSE}
#test that its not all techincal variation from the two batches. It is not. pattern remains the same. 
split <- info %>%
  group_by(Flow.Sort.Strategy) %>%
  do(PCA.res = pca_custom(expnData = Log2.CPM, CDE = .))

split
```

```{r}
unique(info$RNA.to.send..ng.) %>%
  .[order(.)]
```


```{r fig.height=8, fig.width=12}
dat <- split$PCA.res[[1]]$scores
# tiff("TARGET_AML_1031_FlowSorted_Sorted_Leuk_NonLeuk_Samples_withRNAConc_PCAplot.tiff", res=200, units = "in", height = 8, width = 12)

split$PCA.res[[1]]$plot + 
  scale_fill_gradientn(colors=c("lightblue","grey","black"), breaks=c(70,200,300)) + 
  labs(title="PCA Analysis of Leukemic/Non-Leukemic Flow Sorted AML Samples") +
  lims(y=c(-175,100), x=c(-100,300)) +
  # geom_text(aes(label=USI),hjust=-0.25, vjust=-1.5)
  geom_text(data=subset(dat, grepl("PAYGYM|PAYHMK",dat$USI)),
  mapping = aes(x=PC1,y=PC2,label=USI), hjust=-0.15, vjust=-1.1)
  
# dev.off()
```


```{r}
subset(dat2, grepl("PAYKSD|PAYKYZ|PAYKGV|PAYLEK",dat2$USI) & 
                          PC2 > -25 & PC2 < 25)
```

```{r fig.height=8, fig.width=12}
dat2 <- split$PCA.res[[2]]$scores


# tiff("TARGET_AML_1031_FlowSorted_Sorted_CD34_MyeloMon_Lympho_withRNAConc_AddedLabels_PCAplot.tiff", res=200, units = "in", height = 8, width = 12)

split$PCA.res[[2]]$plot + 
  scale_fill_gradientn(colors=c("lightblue","grey","black"), breaks=c(50,200,300)) +
  lims(x=c(-150,250), y=c(-150,150)) 
  # geom_text(aes(label=USI),hjust=-0.15, vjust=-1.2)
  # geom_text(data=subset(dat2, grepl("PAYKSD|PAYKYZ|PAYKGV|PAYLEK",dat2$USI) & 
  #                         PC2 > -25 & PC2 < 25),
  #           mapping = aes(x=PC1,y=PC2,label=USI), hjust=-0.15, vjust=-1.1)
  # geom_text(data=subset(dat, grepl("PAYGYM|PAYHMK",dat$USI)),
  # mapping = aes(x=PC1,y=PC2,label=USI), hjust=-0.15, vjust=-1.1)

# dev.off()
```

Try this next
geom_label_repel(aes(label = Name),
                  box.padding   = 0.35, 
                  point.padding = 0.5,
                  segment.color = 'grey50') +
  theme_classic()


```{r}
table(dat2$Group)
```

#SessionInfo

```{r}
sessionInfo()
```

